Source/IGListAutoSectionController.m

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "IGListAutoSectionController.h"
+
+#import <IGListKit/IGListAssert.h>
+#import <IGListKit/IGListDiffable.h>
+#import <IGListKit/IGListDiff.h>
+#import <IGListKit/IGListBindable.h>
+
+typedef NS_ENUM(NSInteger, IGListAutoSectionState) {
+    IGListAutoSectionStateIdle = 0,
+    IGListAutoSectionStateUpdateQueued,
+    IGListAutoSectionStateUpdateApplied
+};
+
+@interface IGListAutoSectionController()
+
+@property (nonatomic, strong) id object;
+@property (nonatomic, strong) NSArray<id<IGListDiffable>> *viewModels;
+@property (nonatomic, assign) IGListAutoSectionState state;
+
+@end
+
+@implementation IGListAutoSectionController
+
+#pragma mark - Public API
+
+- (void)updateAnimated:(BOOL)animated completion:(void (^)())completion {
+    IGAssertMainThread();
+
+    if (self.state != IGListAutoSectionStateIdle) {
+        return;
+    }
+    self.state = IGListAutoSectionStateUpdateQueued;
+
+    __block IGListIndexSetResult *result = nil;
+    __block NSArray<id<IGListDiffable>> *oldViewModels = nil;
+
+    id<IGListCollectionContext> collectionContext = self.collectionContext;
+
+    [collectionContext performBatchAnimated:YES updates:^{
+        if (self.state != IGListAutoSectionStateUpdateQueued) {
+            return;
+        }
+
+        oldViewModels = self.viewModels;
+        self.viewModels = [self.dataSource viewModelsForObject:self.object];
+        result = IGListDiff(oldViewModels, self.viewModels, IGListDiffEquality);
+
+        [collectionContext deleteInSectionController:self atIndexes:result.deletes];
+        [collectionContext insertInSectionController:self atIndexes:result.inserts];
```

Comment:
Should we support formally `reloading` the item for an updated view-model, rather than only binding a new view model to the existing cell?

One idea is, we add a data source method like:
`- (BOOL)shouldReloadCellForUpdatedViewModel:(id)viewModel atIndex:(NSInteger)index`

And call it for each updated view model to decide whether to reload or just `-bindToViewModel:`.